pkg/cmd/cli/config/store.go

```diff
@@ -0,0 +1,169 @@
+package config
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/clientcmd"
+	clientcmdapi "github.com/GoogleCloudPlatform/kubernetes/pkg/client/clientcmd/api"
+	cmdutil "github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl/cmd/util"
+	"github.com/golang/glog"
+	"github.com/spf13/cobra"
+)
+
+const (
+	OpenShiftConfigPathEnvVar      = "OPENSHIFTCONFIG"
+	OpenShiftConfigFlagName        = "config"
+	OpenShiftConfigFileName        = ".openshiftconfig"
+	OpenShiftConfigHomeDir         = ".config/openshift"
+	OpenShiftConfigHomeFileName    = ".config"
+	OpenShiftConfigHomeDirFileName = OpenShiftConfigHomeDir + "/" + OpenShiftConfigHomeFileName
+
+	KubeConfigPathEnvVar = clientcmd.RecommendedConfigPathEnvVar
+	KubeConfigFileName   = ".kubeconfig"
+	KubeConfigHomeDir    = ".kube"
+
+	fromFlag     = "flag"
+	fromEnvVar   = "envvar"
+	fromLocalDir = "localdir"
+	fromHomeDir  = "homedir"
+
+	fromKube      = "fromkube"
+	fromOpenShift = "fromopenshift"
+)
+
+type ConfigStore struct {
+	Config           *clientcmdapi.Config
+	Path             string
+	providerEngine   string
+	providerLocation string
+}
+
+func (c *ConfigStore) FromFlag() bool {
+	return c.providerLocation == fromFlag
+}
+
+func (c *ConfigStore) FromEnvVar() bool {
+	return c.providerLocation == fromEnvVar
+}
+
+func (c *ConfigStore) FromLocalDir() bool {
+	return c.providerLocation == fromLocalDir
+}
+
+func (c *ConfigStore) FromHomeDir() bool {
+	return c.providerLocation == fromHomeDir
+}
+
+func (c *ConfigStore) FromOpenShift() bool {
+	return c.providerEngine == fromOpenShift
+}
+
+func (c *ConfigStore) FromKube() bool {
+	return c.providerEngine == fromKube
+}
+
+func GetConfigFromDefaultLocations(clientCfg *client.Config, cmd *cobra.Command) (*ConfigStore, error) {
```

Comment:
----- Original Message -----
> > +	return c.providerLocation == fromLocalDir
> > +}
> > +
> > +func (c *ConfigStore) FromHomeDir() bool {
> > +	return c.providerLocation == fromHomeDir
> > +}
> > +
> > +func (c *ConfigStore) FromOpenShift() bool {
> > +	return c.providerEngine == fromOpenShift
> > +}
> > +
> > +func (c *ConfigStore) FromKube() bool {
> > +	return c.providerEngine == fromKube
> > +}
> > +
> > +func GetConfigFromDefaultLocations(clientCfg *client.Config, cmd
> > *cobra.Command) (*ConfigStore, error) {
> 
> If we have this function, why do we need the changes up above to the client
> load order?

The client config loading rules in `loader.go` gets files from all valid locations and merge them all into one set of configs, that will later be used *read-only* by all commands. 

Here we are *managing* a config file, so we target at *one* single file which we need to know the path because we are going to *save* configs to it. It's similar to what is done in `kubectl config` when managing sets of configs (which we may extract from the command in future, refactor and reuse):

https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/kubectl/cmd/config/config.go#L68

> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/992/files#r25139060